Refactor the shutdown

ph opened the pull request:
This PR introduces the changes needed to follow the new shutdown
semantic, it also remove the gelf as a runtime dependencies and move it
as a development dependency since its only used in the tests.

Fixes #17

jsvd commented on spec/inputs/gelf_spec.rb:
```diff
@@ -1,8 +1,22 @@
+Thread.abort_on_exception = true
+describe LogStash::Inputs::Gelf do
+  context "when interrupting the plugin" do
+    let(:port) { "12333" }
```
unless we randomize this we'll probably have issues when jenkins tests the multiple jvms in parallel

ph replied:
good point, changing it.

ph replied:
and wth I used a string here.

purbon replied:
yay! doing ```rand(1024...65532)``` I guess would be the most you need :P

jsvd commented:
there might be some race condition happening here, I only got the following once:

```
logstash-input-gelf (git)-[pr/19] % bundle exec rspec 
Using Accessor#strict_set for specs
Run options: exclude {:redis=>true, :socket=>true, :performance=>true, :couchdb=>true, :elasticsearch=>true, :elasticsearch_secure=>true, :export_cypher=>true, :integration=>true, :windows=>true}
.F

Failures:

  1) LogStash::Inputs::Gelf when interrupting the plugin behaves like an interruptible input plugin #stop returns from run
     Failure/Error: subject.stop
     IOError:
       No such file or directory
     Shared Example Group: "an interruptible input plugin" called from ./spec/inputs/gelf_spec.rb:18
     # ./lib/logstash/inputs/gelf.rb:73:in `stop'
     # /Users/joaoduarte/projects/logstash-devutils/lib/logstash/devutils/rspec/shared_examples.rb:16:in `(root)'
     # /Users/joaoduarte/.rvm/gems/jruby-1.7.19/gems/rspec-wait-0.0.7/lib/rspec/wait.rb:46:in `(root)'

Finished in 1 second (files took 2.66 seconds to load)
2 examples, 1 failure

Failed examples:

rspec /Users/joaoduarte/projects/logstash-devutils/lib/logstash/devutils/rspec/shared_examples.rb:9 # LogStash::Inputs::Gelf when interrupting the plugin behaves like an interruptible input plugin #stop returns from run

Randomized with seed 22953
```

ph commented:
@jsvd looking at it, I guess we have it for a while. ;)

ph commented:
I can reproduce it with your seed. <3

ph commented:
@jsvd it was in a middle of a crash, I have changed the code to use `stoppable_sleep` and added a rescue `IOError` when trying to close the socket, because its totally possible to try to close a socket which is in an inconsistent state.

jsvd commented on lib/logstash/inputs/gelf.rb:
```diff
@@ -60,42 +61,32 @@ def run(output_queue)
-    end
-    finished
+  def stop
+    super
```
super is no longer needed

jsvd commented on logstash-input-gelf.gemspec:
```diff
@@ -23,9 +23,10 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency "gelfd", ["0.2.0"]                 #(Apache 2.0 license)
-  s.add_runtime_dependency "gelf", ["1.3.2"]                  #(MIT license)
   s.add_runtime_dependency 'logstash-codec-plain'
+  s.add_runtime_dependency "stud"
```
I'd suggest adding a ~> 0.0.22 or something similar

jsvd commented on spec/inputs/gelf_spec.rb:
```diff
@@ -1,8 +1,22 @@
+Thread.abort_on_exception = true
+describe LogStash::Inputs::Gelf do
+  context "when interrupting the plugin" do
+    let(:port) { rand(1024..65535) }
```
until we tie the random seed to rspec seed in devutils can you do it here?

ph replied:
Er.. No I'll fix it in in devutils, many plugins already use some sort of randomness.
I was expecting rspec to actually monkey patch the seed, but after verification this is not the case.

jsvd replied:
Some plugins might not want this for some reason, so the right way is to make this configurable somehow. The problem is that Rspec.configuration block is global :(

ph replied:
Okay, this is what I propose.

- I'll change rand for Flores::Random
- Fix flores to use the rspec seed, which is okay in that case since its an explicitl rspec helper I assume it to have repeatable run.

If people want pure random with no way to repeat it, they can fallback to `Kernel.rand` and manage the seed themself. What do you think? I think this work isn't part of this PR.

ph replied:
Concerning the `rspec.configuration` block both `logstash` and devutils should provide common default for tests including seed propagation.

ph commented:
Weird, non deterministic errors are back, investigating.

ph commented:
ready for another round of review @jsvd

jsvd commented on spec/inputs/gelf_spec.rb:
```diff
@@ -1,8 +1,25 @@
+require "flores/random"
 
-describe "inputs/gelf" do
+Thread.abort_on_exception = true
```
is this really needed?

jsvd commented:
Minor comment, otherwise LGTM. squash and merge :sheep: it
